README.md
# typed-ip

A Typescript library with zero dependencies to parse and manipulate IP addresses.

This project is **NOT** ready for production right now, you may expect changes to the public interfaces.


## Installing

```shell
$ yarn install typed-ip
```


## Building

Before starting, install the dependencies:

```shell
$ yarn install
```


To build the project:

```shell
$ npm run build
```


To create an installable package:

```shell
$ npm run pack
```


To run the tests directly in Typescript (quickier, better while developing):
```shell
$ npm run test
```


To run the tests in Javascript (includes coverage):

```shell
$ npm run test-js
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
* Add node v10 to Travis CI.

## [0.3.3]
### Changed
* Renamed the package from `tsipaddr` to `typed-ip`.

## [0.3.2]
### Changed
* Change README from reStructuredText to Markdown for compatibility with npmjs.com

## [0.3.1]
### Added
* Add `chai` assert library.
* Add Travis CI integration.
### Changed
* Make tests single-assert - it's easier for quickly debugging errors.
* Upgrade dependencies.
### Removed
* Remove unnecessary `Network6` tests.
### Fixed
* Fix TS2564 for Typescript >=2.7
* Mark as hidden some forgotten private functions.

## [0.3.0] - 2018-04-21
### Added
- Add `getDecimal` function to `Address4`.
- Add `Address6` and `Network6` classes, initial IPv6 support.
- Add `getFullString` and `getRfc5952` to `Address6`.
### Changed
- Small optimization in broadcast address calculation.

## [0.2.0] - 2017-11-16
### Added
- This changelog :)
- ES3 compatibility.
- Add public documentation; generate it with `npm run doc`.
- Add tests for exceptions.
- Add the internal library `ByteContainer`.
### Changed
- Replaced properties accessors with `get` functions, as Typescript cannot
  generate ES3 code with them.
- All public methods are now explicitly typed.
- Completely remove `ByteArray`, now `ReadonlyArray<number>` is used instead for public interfaces.
- Renamed `getHostmask` with `getWildcard` along with all the internal names and comments.
### Fixed
- Fixed network address calculation.

## [0.1.0] - 2017-11-01
### Added
- First release.


tests/address4.ts
import { assert } from "chai";

import {
    Address,
    Address4,
    AddressValueError,
} from "../src";


interface AddressInfo {
    address: string;
    octets: number[];
    decimal: number;
}

interface ArithmeticInfo {
    octets: number[];
    quantity: number[];
    result: number[];
}

interface ComparisonInfo {
    left: number[];
    right: number[];
}

/* tslint:disable:object-literal-sort-keys */
const IPV4_VALID: AddressInfo[] = [
    { address: "0.0.0.0", octets: [0, 0, 0, 0], decimal: 0 },
    { address: "127.0.0.1", octets: [127, 0, 0, 1], decimal: 2130706433 },
    { address: "192.168.0.1", octets: [192, 168, 0, 1], decimal: 3232235521 },
    { address: "255.255.255.0", octets: [255, 255, 255, 0], decimal: 4294967040 },
    { address: "255.255.255.255", octets: [255, 255, 255, 255], decimal: 4294967295 },
];

const INVALID_STRINGS = [
    "256.0.0.0",
    "192.0.-1.0",
    "192.168.0",
    "192.168.0.1.0",
    "192.a.0.1",
];

const INVALID_OCTETS = [
    [-1],
    [256, 0, 0, 0],
];

const ADD_VALID: ArithmeticInfo[] = [
    { octets: [127, 0, 0, 1], quantity: [1], result: [127, 0, 0, 2] },
    { octets: [127, 0, 10, 255], quantity: [1], result: [127, 0, 11, 0] },
    { octets: [127, 10, 255, 255], quantity: [2], result: [127, 11, 0, 1] },
    { octets: [0, 255, 255, 254], quantity: [3], result: [1, 0, 0, 1] },
    { octets: [0, 255, 255, 255], quantity: [1, 255], result: [1, 0, 1, 254] },
];

const SUBSTRACT_VALID: ArithmeticInfo[] = [
    { octets: [127, 0, 0, 1], quantity: [1], result: [127, 0, 0, 0] },
    { octets: [127, 0, 0, 1], quantity: [2], result: [126, 255, 255, 255] },
];

const EQUAL_VALID: ComparisonInfo[] = [
    { left: [127, 0, 0, 1], right: [127, 0, 0, 1] },
    { left: [192, 168, 0, 1], right: [192, 168, 0, 1] },
];

const NOT_EQUAL_VALID: ComparisonInfo[] = [
    { left: [127, 0, 0, 1], right: [127, 0, 0, 0] },
    { left: [127, 0, 0, 0], right: [126, 0, 0, 0] },
];

const GREATER_THAN_VALID: ComparisonInfo[] = [
    { left: [127, 0, 0, 1], right: [127, 0, 0, 0] },
    { left: [127, 0, 0, 0], right: [126, 255, 255, 255] },
];

const LESSER_THAN_VALID: ComparisonInfo[] = [
    { left: [127, 0, 0, 1], right: [127, 0, 0, 2] },
    { left: [126, 0, 255, 255], right: [126, 255, 0, 0] },
];


function describeTestAddress(describeSuffix: string, addressConstructor: (data: AddressInfo) => Address4) {
    describe(`public interface ${describeSuffix}`, function() {
        it("should have the correct type", function() {
            assert.instanceOf(addressConstructor(IPV4_VALID[0]), Address);
        });
        IPV4_VALID.forEach((input) => {
            it(`should match string address ${input.address}`, function() {
                assert.equal(addressConstructor(input).getIpString(), input.address);
            });
            it(`should match decimal address ${input.address}`, function() {
                assert.equal(addressConstructor(input).getDecimal(), input.decimal);
            });
            it(`should match octets ${input.address}`, function() {
                assert.deepEqual(addressConstructor(input).getOctets(), input.octets);
            });
        });
    });
}


describe("Address4", function() {
    describe("constructor from string", function() {
        IPV4_VALID.forEach((input) => {
            it(`should accept valid string value ${input.address}`, function() {
                assert.ok(new Address4(input.address));
            });
        });
        INVALID_STRINGS.forEach((input) => {
            it(`should reject invalid string value ${input}`, function() {
                assert.throws(() => new Address4(input), AddressValueError);
            });
        });
    });
    describe("constructor from octets", function() {
        IPV4_VALID.forEach((input) => {
            it(`should accept valid octets value ${input.octets}`, function() {
                assert.ok(new Address4(input.octets));
            });
        });
        INVALID_OCTETS.forEach((input) => {
            it(`should reject invalid octets value ${input}`, function() {
                assert.throws(() => new Address4(input), AddressValueError);
            });
        });
    });
    describe("arithmetic", function() {
        ADD_VALID.forEach((input) => {
            it(`should add ${input.quantity} to ${input.octets}`, function() {
                assert.deepEqual(
                    new Address4(input.octets).add(input.quantity).getOctets(),
                    input.result,
                );
            });
        });
        SUBSTRACT_VALID.forEach((input) => {
            it(`should substract ${input.quantity} to ${input.octets}`, function() {
                assert.deepEqual(
                    new Address4(input.octets).substract(input.quantity).getOctets(),
                    input.result,
                );
            });
        });
    });
    describe("comparisons", function() {
        NOT_EQUAL_VALID.forEach((input) => {
            it(`should not compare equal ${input.left} and ${input.right}`, function() {
                    assert.isFalse(new Address4(input.left).eq(new Address4(input.right)));
            });
            it(`should compare not equal ${input.left} and ${input.right}`, function() {
                    assert.isTrue(new Address4(input.left).ne(new Address4(input.right)));
            });
        });
        EQUAL_VALID.forEach((input) => {
            it(`should compare equal ${input.left} and ${input.right}`, function() {
                    assert.isTrue(new Address4(input.left).eq(new Address4(input.right)));
            });
            it(`should not compare not equal ${input.left} and ${input.right}`, function() {
                    assert.isFalse(new Address4(input.left).ne(new Address4(input.right)));
            });
            it(`should not compare greater ${input.left} than ${input.right}`, function() {
                assert.isFalse(new Address4(input.left).gt(new Address4(input.right)));
            });
            it(`should compare greater or equal ${input.left} than ${input.right}`, function() {
                assert.isTrue(new Address4(input.left).ge(new Address4(input.right)));
            });
            it(`should not compare lesser ${input.left} than ${input.right}`, function() {
                assert.isFalse(new Address4(input.left).lt(new Address4(input.right)));
            });
            it(`should compare lesser or equal ${input.left} than ${input.right}`, function() {
                assert.isTrue(new Address4(input.left).le(new Address4(input.right)));
            });
        });
        GREATER_THAN_VALID.forEach((input) => {
            it(`should not compare equal ${input.left} and ${input.right}`, function() {
                    assert.isFalse(new Address4(input.left).eq(new Address4(input.right)));
            });
            it(`should compare not equal ${input.left} and ${input.right}`, function() {
                    assert.isTrue(new Address4(input.left).ne(new Address4(input.right)));
            });
            it(`should compare greater ${input.left} than ${input.right}`, function() {
                assert.isTrue(new Address4(input.left).gt(new Address4(input.right)));
            });
            it(`should compare greater or equal ${input.left} than ${input.right}`, function() {
                assert.isTrue(new Address4(input.left).ge(new Address4(input.right)));
            });
            it(`should not compare lesser ${input.left} than ${input.right}`, function() {
                assert.isFalse(new Address4(input.left).lt(new Address4(input.right)));
            });
            it(`should not compare lesser or equal ${input.left} than ${input.right}`, function() {
                assert.isFalse(new Address4(input.left).le(new Address4(input.right)));
            });
        });
        LESSER_THAN_VALID.forEach((input) => {
            it(`should not compare equal ${input.left} and ${input.right}`, function() {
                    assert.isFalse(new Address4(input.left).eq(new Address4(input.right)));
            });
            it(`should compare not equal ${input.left} and ${input.right}`, function() {
                    assert.isTrue(new Address4(input.left).ne(new Address4(input.right)));
            });
            it(`should not compare greater ${input.left} than ${input.right}`, function() {
                assert.isFalse(new Address4(input.left).gt(new Address4(input.right)));
            });
            it(`should not compare greater or equal ${input.left} than ${input.right}`, function() {
                assert.isFalse(new Address4(input.left).ge(new Address4(input.right)));
            });
            it(`should compare lesser ${input.left} than ${input.right}`, function() {
                assert.isTrue(new Address4(input.left).lt(new Address4(input.right)));
            });
            it(`should compare lesser or equal ${input.left} than ${input.right}`, function() {
                assert.isTrue(new Address4(input.left).le(new Address4(input.right)));
            });
        });
    });
    describeTestAddress("from string", (data: AddressInfo) => new Address4(data.address));
    describeTestAddress("from octets", (data: AddressInfo) => new Address4(data.octets));
});


tests/address6.ts
import { assert } from "chai";

import {
    Address,
    Address6,
    AddressValueError,
} from "../src";

interface AddressInfo {
    address: string;
    octets: number[];
    addressFull: string;
    addressRfc5952: string;
}


/* tslint:disable:object-literal-sort-keys */
const IPV6_VALID: AddressInfo[] = [
    {
        address: "::127.0.0.1",
        octets: [0, 0, 0, 0, 0, 0, 0, 0,
                 0, 0, 0, 0, 127, 0, 0, 1],
        addressFull: "0000:0000:0000:0000:0000:0000:7f00:0001",
        addressRfc5952: "::7f00:1",
    },
    {
        address: "::7f00:1",
        octets: [0, 0, 0, 0, 0, 0, 0, 0,
                 0, 0, 0, 0, 127, 0, 0, 1],
        addressFull: "0000:0000:0000:0000:0000:0000:7f00:0001",
        addressRfc5952: "::7f00:1",
    },
    {
        address: "2001:db8:0:0:1:0:0:1",
        octets: [32, 1, 13, 184, 0, 0, 0, 0,
                 0, 1, 0, 0, 0, 0, 0, 1],
        addressFull: "2001:0db8:0000:0000:0001:0000:0000:0001",
        addressRfc5952: "2001:db8::1:0:0:1",
    },
    {
        address: "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff",
        octets: [255, 255, 255, 255, 255, 255, 255, 255,
                 255, 255, 255, 255, 255, 255, 255, 255],
        addressFull: "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff",
        addressRfc5952: "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff",
    },
    {
        address: "1:0:ffff:ffff:ffff:ffff:ffff:ffff",
        octets: [0, 1, 0, 0, 255, 255, 255, 255,
                 255, 255, 255, 255, 255, 255, 255, 255],
        addressFull: "0001:0000:ffff:ffff:ffff:ffff:ffff:ffff",
        addressRfc5952: "1:0:ffff:ffff:ffff:ffff:ffff:ffff",
    },
    {
        address: "::ffff:ffff:ffff:ffff:ffff:ffff",
        octets: [0, 0, 0, 0, 255, 255, 255, 255,
                 255, 255, 255, 255, 255, 255, 255, 255],
        addressFull: "0000:0000:ffff:ffff:ffff:ffff:ffff:ffff",
        addressRfc5952: "::ffff:ffff:ffff:ffff:ffff:ffff",
    },
    {
        address: "ffff:ffff:ffff:ffff:ffff:ffff::",
        octets: [255, 255, 255, 255, 255, 255, 255, 255,
                 255, 255, 255, 255, 0, 0, 0, 0],
        addressFull: "ffff:ffff:ffff:ffff:ffff:ffff:0000:0000",
        addressRfc5952: "ffff:ffff:ffff:ffff:ffff:ffff::",
    },
];


function describeTestAddress(describeSuffix: string, addressConstructor: (data: AddressInfo) => Address6) {
    describe(`public interface ${describeSuffix}`, function() {
        it("should have the correct type", function() {
            assert.instanceOf(addressConstructor(IPV6_VALID[0]), Address);
        });
        IPV6_VALID.forEach((input) => {
            it(`${input.address} should match octets ${input.octets}`, function() {
                assert.deepEqual(addressConstructor(input).getOctets(), input.octets);
            });
            it(`${input.address} should match full string address ${input.addressFull}`, function() {
                assert.deepEqual(addressConstructor(input).getFullString(), input.addressFull);
            });
            it(`${input.address} should match RFC 5952 string address ${input.addressRfc5952}`, function() {
                assert.deepEqual(addressConstructor(input).getRfc5952(), input.addressRfc5952);
            });
        });
    });
}


describe("Address6", function() {
    describeTestAddress("from string", (data: AddressInfo) => new Address6(data.address));
    describeTestAddress("from octets", (data: AddressInfo) => new Address6(data.octets));
    describeTestAddress("from full address", (data: AddressInfo) => new Address6(data.addressFull));
    describeTestAddress("from RFC 5952 address", (data: AddressInfo) => new Address6(data.addressRfc5952));
});


tests/network4.ts
import { assert } from "chai";
import {
    Address4,
    Network,
    Network4,
} from "../src";


interface NetworkInfo {
    inputPrefix: string;
    inputAddress: string;
    address: number[];
    prefix: number;
    netmask: number[];
    wildcard: number[];
    broadcast: number[];
    numAddresses: number;
    firstAddress: number[];
    lastAddress: number[];
    beforeFirstAddress: number[];
    afterLastAddress: number[];
}

/* tslint:disable:object-literal-sort-keys */
const IPV4_NETWORKS_VALID: NetworkInfo[] = [{
    inputAddress: "192.168.0.0/255.255.255.0",
    inputPrefix: "192.168.0.0/24",
    address: [192, 168, 0, 0],
    prefix: 24,
    netmask: [255, 255, 255, 0],
    wildcard: [0, 0, 0, 255],
    broadcast: [192, 168, 0, 255],
    numAddresses: 256,
    firstAddress: [192, 168, 0, 1],
    lastAddress: [192, 168, 0, 254],
    beforeFirstAddress: [192, 167, 255, 255],
    afterLastAddress: [192, 168, 1, 0],
}, {
    inputAddress: "10.0.0.0/255.0.0.0",
    inputPrefix: "10.0.0.0/8",
    address: [10, 0, 0, 0],
    prefix: 8,
    netmask: [255, 0, 0, 0],
    wildcard: [0, 255, 255, 255],
    broadcast: [10, 255, 255, 255],
    numAddresses: 16777216,
    firstAddress: [10, 0, 0, 1],
    lastAddress: [10, 255, 255, 254],
    beforeFirstAddress: [9, 255, 255, 255],
    afterLastAddress: [11, 0, 0, 0],
}, {
    inputAddress: "148.56.0.0/255.255.240.0",
    inputPrefix: "148.56.0.0/20",
    address: [148, 56, 0, 0],
    prefix: 20,
    netmask: [255, 255, 240, 0],
    wildcard: [0, 0, 15, 255],
    broadcast: [148, 56, 15, 255],
    numAddresses: 4096,
    firstAddress: [148, 56, 0, 1],
    lastAddress: [148, 56, 15, 254],
    beforeFirstAddress: [148, 55, 0, 0],
    afterLastAddress: [148, 56, 16, 0],
}, {
    inputAddress: "192.168.0.40/255.255.248.0",
    inputPrefix: "192.168.0.40/21",
    address: [192, 168, 0, 0],
    prefix: 21,
    netmask: [255, 255, 248, 0],
    wildcard: [0, 0, 7, 255],
    broadcast: [192, 168, 7, 255],
    numAddresses: 2048,
    firstAddress: [192, 168, 0, 1],
    lastAddress: [192, 168, 7, 254],
    beforeFirstAddress: [192, 167, 255, 255],
    afterLastAddress: [192, 168, 8, 0],
}];


function describeTestNetwork(describeSuffix: string, inputAccessor: (data: NetworkInfo) => string) {
    describe(`public interface ${describeSuffix}`, function() {
        it("should have the Network type", function() {
            assert.instanceOf(new Network4(inputAccessor(IPV4_NETWORKS_VALID[0])), Network);
        });
        it("address should have the Address4 type", function() {
            assert.instanceOf(new Network4(inputAccessor(IPV4_NETWORKS_VALID[0])).getAddress(), Address4);
        });
        it("netmask should have the Address4 type", function() {
            assert.instanceOf(new Network4(inputAccessor(IPV4_NETWORKS_VALID[0])).getNetmask(), Address4);
        });
        it("wildcard should have the Address4 type", function() {
            assert.instanceOf(new Network4(inputAccessor(IPV4_NETWORKS_VALID[0])).getWildcard(), Address4);
        });
        it("broadcast should have the Address4 type", function() {
            assert.instanceOf(new Network4(inputAccessor(IPV4_NETWORKS_VALID[0])).getBroadcast(), Address4);
        });
        IPV4_NETWORKS_VALID.forEach((input) => {
            const inputString = inputAccessor(input);
            it(`${inputString} should match network address ${input.address}`, function() {
                assert.deepEqual(new Network4(inputString).getAddress().getOctets(), input.address);
            });
            it(`${inputString} should match netmask address ${input.netmask}`, function() {
                assert.deepEqual(new Network4(inputString).getNetmask().getOctets(), input.netmask);
            });
            it(`${inputString} should match wildcard address ${input.wildcard}`, function() {
                assert.deepEqual(new Network4(inputString).getWildcard().getOctets(), input.wildcard);
            });
            it(`${inputString} should match broadcast address ${input.broadcast}`, function() {
                assert.deepEqual(new Network4(inputString).getBroadcast().getOctets(), input.broadcast);
            });
            it(`${inputString} should match prefix ${input.prefix}`, function() {
                assert.equal(new Network4(inputString).getPrefix(), input.prefix);
            });
            it(`${inputString} should match number of addresses ${input.numAddresses}`, function() {
                assert.equal(new Network4(inputString).getNumAddresses(), input.numAddresses);
            });
            it(`${inputString} should start with the first address ${input.firstAddress}`, function() {
                assert.deepEqual(new Network4(inputString).hosts().next().value.getOctets(), input.firstAddress);
            });
            it(`${inputString} should end with the last address ${input.lastAddress}`, function() {
                assert.isTrue(new Network4(inputString).hosts(new Address4(input.lastAddress)).next().done);
            });
            it(`${inputString} should contains the first address ${input.firstAddress}`, function() {
                assert.isTrue(new Network4(inputString).contains(new Address4(input.firstAddress)));
            });
            it(`${inputString} should contains the last address ${input.lastAddress}`, function() {
                assert.isTrue(new Network4(inputString).contains(new Address4(input.lastAddress)));
            });
            it(`${inputString} should not contains before the first address ${input.beforeFirstAddress}`, function() {
                assert.isFalse(new Network4(inputString).contains(new Address4(input.beforeFirstAddress)));
            });
            it(`${inputString} should not contains after the last address ${input.afterLastAddress}`, function() {
                assert.isFalse(new Network4(inputString).contains(new Address4(input.afterLastAddress)));
            });
        });
    });

}


describe("Network4", function() {
    describe("constructor", function() {
        IPV4_NETWORKS_VALID.forEach((input) => {
            it(`should accept valid prefix value ${input.inputPrefix}`, function() {
                assert.isOk(new Network4(input.inputPrefix));
            });
            it(`should accept valid netmask value ${input.inputAddress}`, function() {
                assert.isOk(new Network4(input.inputAddress));
            });
        });
    });
    describeTestNetwork("from prefix", (data: NetworkInfo) => data.inputPrefix);
    describeTestNetwork("from address", (data: NetworkInfo) => data.inputAddress);
});


src/address4.ts
import {
    AddressValueError,
} from './exceptions';

import {
    DECIMAL_DIGITS,
    NETMASK_OCTETS,
    IPV4_LENGTH,
    IPV4_BYTES,
} from './constants';

import {
    Address,
    Network,
} from './address';



export class Address4 extends Address {
    private isValidOctet(octet: number) {
        return octet >= 0 && octet <= 255;
    }

    protected getStringFromOctets(octets: Uint8Array) {
        return octets.join(".");
    }

    protected getOctetsFromArray(address: ReadonlyArray<number>) {
        if (address.length > IPV4_BYTES) {
            throw new AddressValueError(address);
        }
        if (!address.every((octet) => this.isValidOctet(octet))) {
            throw new AddressValueError(address);
        }
        if (address.length < IPV4_BYTES) {
            address = Array(IPV4_BYTES - address.length).fill(0x00).concat(address);
        }
        return new Uint8Array(address);
    }

    protected getOctetsFromString(address: string) {
        const addressSplitted = address.split(".");
        if (addressSplitted.length !== IPV4_BYTES) {
            throw new AddressValueError(address);
        }

        const octets = new Uint8Array(addressSplitted.map(
            (octet) => {
                if (!Array.from(octet).every((ch) => DECIMAL_DIGITS.includes(ch))) {
                    throw new AddressValueError(address as string);
                }

                const numberOctet = parseInt(octet, 10);
                if (!this.isValidOctet(numberOctet)) {
                    throw new AddressValueError(address as string);
                }

                return numberOctet;
            },
        ));
        return octets;
    }
}


export class Network4 extends Network<Address4>{
    public constructor(input: string) {
        const [subaddress, subnetmask, nothing] = input.split("/");
        if (typeof nothing !== "undefined") {
            throw new AddressValueError(input);
        }

        if (typeof subnetmask === "undefined") {
            throw new AddressValueError(input);
        }

        const address = new Address4(subaddress);
        if (Array.from(subnetmask).every((ch) => DECIMAL_DIGITS.includes(ch))) {
            const prefix = parseInt(subnetmask, 10);
            if (prefix < 0 || prefix > IPV4_LENGTH) {
                throw new AddressValueError(subnetmask);
            }
            const fullBytes = Math.floor(prefix / 8);
            const partialByte = prefix % 8;
            const bytes: number[] = Array(IPV4_BYTES);
            if (fullBytes > 0) {
                bytes.fill(0xff, 0, fullBytes);
            }
            if (partialByte > 0) {
                bytes[IPV4_BYTES - fullBytes] = NETMASK_OCTETS[partialByte];
            }
            var netmask = new Address4(bytes);
        } else {
            var netmask = new Address4(subnetmask);
            if (!netmask.octets.every((octet) => NETMASK_OCTETS.includes(octet))) {
                throw new AddressValueError(subnetmask);
            }
        }
        super(address, netmask, Address4);
    }
}


src/pyipaddr.ts
const DECIMAL_DIGITS = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
const NETMASK_OCTETS = [
    0,
    128,
    192,
    224,
    240,
    248,
    252,
    254,
    255,
];


export const IPV4_LENGTH = 32;
export const IPV4_BYTES = 4;


export class AddressValueError extends Error {
    public constructor(address: string | ReadonlyArray<number>) {
        super("Invalid address: " + address);
        // Needed for Typescript 2.1 and following
        // See https://github.com/Microsoft/TypeScript-wiki/blob/77a2a18a6e7bf8599bbe693243b1b9eb3044bfda/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
        Object.setPrototypeOf(this, AddressValueError.prototype);
    }
}


export class NetmaskValueError extends Error {
    public constructor(netmask: string | ReadonlyArray<number>) {
        super("Invalid netmask: " + netmask);
        // Needed for Typescript 2.1 and following
        // See https://github.com/Microsoft/TypeScript-wiki/blob/77a2a18a6e7bf8599bbe693243b1b9eb3044bfda/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
        Object.setPrototypeOf(this, NetmaskValueError.prototype);
    }
}


const enum comparison {
    Greater, Lesser, Equal,
}


type comparator<T> = (left: T, right: T) => comparison;


function compareArrays<T>(
    left: ArrayLike<T>,
    right: ArrayLike<T>,
    func: comparator<T>,
): comparison {
    if (left.length > right.length) {
        return comparison.Greater;
    }

    if (right.length > left.length) {
        return comparison.Lesser;
    }

    for (let index = 0; index < left.length; index++) {
        const result = func(left[index], right[index]);
        if (result !== comparison.Equal) {
            return result;
        }
    }

    return comparison.Equal;
}


function compareNumbers(left: number, right: number) {
    if (left > right) {
        return comparison.Greater;
    }

    if (right > left) {
        return comparison.Lesser;
    }

    return comparison.Equal;
}


function compareByteArrays(left: Uint8Array, right: Uint8Array) {
    return compareArrays(left, right, compareNumbers);
}


function addByteArrays(left: Uint8Array, right: Uint8Array) {
    if (right.length > left.length) {
        throw new Error("Unsupported operation: right cannot have more elements than left");
    }
    if (right.length < left.length) {
        right = new Uint8Array(Array(left.length - right.length).fill(0x00).concat(right));
    }
    const copy = new Uint8Array(left);
    let carry = 0;
    for (let index = 0; index < right.length; index++) {
        const rightIndex = right.length - index - 1;
        if (right[rightIndex] + carry === 0) {
            continue;
        }

        const leftIndex = left.length - index - 1;
        const sum = left[leftIndex] + right[rightIndex] + carry;
        if (sum > 255) {
            copy[leftIndex] = sum - 256;
            carry = 1;
        } else {
            copy[leftIndex] = sum;
            carry = 0;
        }
    }

    if (carry > 0) {
        throw new Error("Overflow");
    }

    return copy;
}


function substractByteArrays(left: Uint8Array, right: Uint8Array) {
    if (right.length > left.length) {
        throw new Error("Unsupported operation: right cannot have more elements than left");
    }
    if (right.length < left.length) {
        right = new Uint8Array(Array(left.length - right.length).fill(0x00).concat(right));
    }
    const copy = new Uint8Array(left);
    let carry = 0;
    for (let index = 0; index < right.length; index++) {
        const rightIndex = right.length - index - 1;
        if (right[rightIndex] + carry === 0) {
            continue;
        }

        const leftIndex = left.length - index - 1;
        const subs = left[leftIndex] - right[rightIndex] - carry;
        if (subs < 0) {
            copy[leftIndex] = subs + 256;
            carry = 1;
        } else {
            copy[leftIndex] = subs;
            carry = 0;
        }
    }

    if (carry > 0) {
        throw new Error("Underflow");
    }

    return copy;
}


export class IPv4Address {
    private _octets: Uint8Array;
    private _ipString: string;

    public constructor(address: string | ReadonlyArray<number> | Uint8Array) {
        if (address instanceof Uint8Array) {
            this._octets = new Uint8Array(address);
        } else if (address instanceof Array) {
            if (address.length > IPV4_BYTES) {
                throw new AddressValueError(address);
            }
            if (!address.every((byte) => byte > -1 && byte < 256)) {
                throw new AddressValueError(address);
            }
            if (address.length < IPV4_BYTES) {
                address = Array(IPV4_BYTES - address.length).fill(0x00).concat(address);
            }
            this._octets = new Uint8Array(address);
        } else if (typeof address === "string") {
            const addressSplitted = address.split(".");
            if (addressSplitted.length !== IPV4_BYTES) {
                throw new AddressValueError(address);
            }

            this._octets = new Uint8Array(addressSplitted.map(
                (octet) => {
                    if (!Array.from(octet).every((ch) => DECIMAL_DIGITS.includes(ch))) {
                        throw new AddressValueError(address as string);
                    }

                    const numberOctet = parseInt(octet, 10);
                    if (numberOctet < 0 || numberOctet > 255) {
                        throw new AddressValueError(address as string);
                    }

                    return numberOctet;
                },
            ));
        }
    }

    get ipString() {
        if (typeof this._ipString === "undefined") {
            this._ipString = this._octets.join(".");
        }

        return this._ipString;
    }

    get octets() {
        return this._octets;
    }

    public eq(other: IPv4Address) {
        return compareByteArrays(this.octets, other.octets) === comparison.Equal;
    }

    public ne(other: IPv4Address) {
        return compareByteArrays(this.octets, other.octets) !== comparison.Equal;
    }

    public gt(other: IPv4Address) {
        return compareByteArrays(this.octets, other.octets) === comparison.Greater;
    }

    public lt(other: IPv4Address) {
        return compareByteArrays(this.octets, other.octets) === comparison.Lesser;
    }

    public ge(other: IPv4Address) {
        const result = compareByteArrays(this.octets, other.octets);
        return result === comparison.Greater || result === comparison.Equal;
    }

    public le(other: IPv4Address) {
        const result = compareByteArrays(this.octets, other.octets);
        return result === comparison.Lesser || result === comparison.Equal;
    }

    public add(amount: ReadonlyArray<number>) {
        return new IPv4Address(addByteArrays(this.octets, new Uint8Array(amount)));
    }

    public substract(amount: ReadonlyArray<number>) {
        return new IPv4Address(substractByteArrays(this.octets, new Uint8Array(amount)));
    }

    public toString() {
        return this.ipString;
    }
}


export class IPv4Network {
    private _address: IPv4Address;
    private _netmask: IPv4Address;
    private _prefix: number;
    private _hostmask: IPv4Address;
    private _broadcast: IPv4Address;
    private _numAddresses: number;

    public constructor(address: string) {
        const [subaddress, subnetmask, nothing] = address.split("/");
        if (typeof nothing !== "undefined") {
            throw new AddressValueError(address);
        }

        if (typeof subnetmask === "undefined") {
            throw new NetmaskValueError(address);
        }

        this._address = new IPv4Address(subaddress);
        if (Array.from(subnetmask).every((ch) => DECIMAL_DIGITS.includes(ch))) {
            const netmask = parseInt(subnetmask, 10);
            if (netmask < 0 || netmask > IPV4_LENGTH) {
                throw new NetmaskValueError(subnetmask);
            }
            const fullBytes = Math.floor(netmask / 8);
            const partialByte = netmask % 8;
            const bytes: number[] = Array(IPV4_BYTES);
            if (fullBytes > 0) {
                bytes.fill(0xff, 0, fullBytes);
            }
            if (partialByte > 0) {
                bytes[IPV4_BYTES - fullBytes] = NETMASK_OCTETS[partialByte];
            }
            this._netmask = new IPv4Address(bytes);
        } else {
            this._netmask = new IPv4Address(subnetmask);
            if (!this._netmask.octets.every((octet) => NETMASK_OCTETS.includes(octet))) {
                throw new NetmaskValueError(subnetmask);
            }
        }
    }

    get address() {
        return this._address;
    }

    get netmask() {
        return this._netmask;
    }

    get prefix() {
        if (typeof this._prefix === "undefined") {
            this._prefix = this.netmask.octets.reduce(
                (prefix, octet) => prefix + NETMASK_OCTETS.indexOf(octet),
                0,
            );
        }

        return this._prefix;
    }

    get hostmask() {
        if (typeof this._hostmask === "undefined") {
            this._hostmask = new IPv4Address(this.netmask.octets.map((octet) => ~octet & 0xff).join("."));
        }

        return this._hostmask;
    }

    get broadcast() {
        if (typeof this._broadcast === "undefined") {
            this._broadcast = new IPv4Address([
                this.address.octets[0] | this.hostmask.octets[0],
                this.address.octets[1] | this.hostmask.octets[1],
                this.address.octets[2] | this.hostmask.octets[2],
                this.address.octets[3] | this.hostmask.octets[3],
            ].join("."));
        }

        return this._broadcast;
    }

    get numAddresses() {
        if (typeof this._numAddresses === "undefined") {
            this._numAddresses = this.hostmask.octets.reduce(
                (total, octet, index) => total + octet * (256 ** (this.hostmask.octets.length - index - 1)),
                    1,
            );
        }

        return this._numAddresses;
    }

    public * hosts(start?: IPv4Address) {
        if (!start) {
            start = this.address;
        }

        let address = start;
        const maxAddress = this.broadcast.substract([1]);
        while (address.lt(maxAddress)) {
            address = address.add([1]);
            yield address;
        }
    }

    public contains(other: IPv4Address) {
        return (this.broadcast.ge(other) && this.address.le(other));
    }
}
